File: service/service/trainee/trainee-api/src/main/java/bs/service/trainee/api/service/TraineeAttendanceService.java
package bs.service.trainee.api.service;

import bs.lib.common.model.generated.LookupResultSet;
import bs.lib.common.model.generated.NewRecordVTO;
import bs.lib.sql.db.adapter.model.generated.OrderDirections;
import bs.service.trainee.model.enums.TraineeAttendanceStatus;
import bs.service.trainee.model.generated.*;

import java.time.LocalDate;
import java.util.List;

/**
 * Service interface for managing trainee attendance
 */
public interface TraineeAttendanceService {

    // ==================== Basic CRUD Operations ====================

    /**
     * Create a new trainee attendance record
     *
     * @param traineeAttendanceDTO the attendance data
     * @return NewRecordVTO containing the created record ID
     */
    NewRecordVTO createTraineeAttendance(TraineeAttendanceDTO traineeAttendanceDTO);

    /**
     * Update an existing trainee attendance record
     *
     * @param attendanceId the ID of the attendance record
     * @param traineeAttendanceDTO the updated attendance data
     * @return NewRecordVTO containing the updated record ID
     */
    NewRecordVTO updateTraineeAttendance(Integer attendanceId, TraineeAttendanceDTO traineeAttendanceDTO);

    /**
     * Soft delete a trainee attendance record
     *
     * @param attendanceId the ID of the attendance record to delete
     */
    void deleteTraineeAttendanceById(Integer attendanceId);

    /**
     * Get a trainee attendance record by ID
     *
     * @param attendanceId the ID of the attendance record
     * @return TraineeAttendanceVTO containing the attendance details
     */
    TraineeAttendanceVTO getTraineeAttendanceById(Integer attendanceId);


    // ==================== Query Operations ====================

    TraineeAttendanceResultSet getAllTraineeAttendances(TraineeAttendanceStatus status,String sessionDay, String traineeNationalId, Integer traineeId, Integer courseId, Integer courseSessionId, String quickSearch, String checkInFrom, String checkInTo, String checkOutFrom, String checkOutTo, LocalDate fromDate, LocalDate toDate, Integer pageNum, Integer pageSize, OrderDirections orderDir, String orderBy);

    /**
     * Get daily attendance report for a specific date
     *
     * @param attendanceDate the date to generate report for
     * @return DailyAttendanceReport containing attendance statistics
     */
    DailyAttendanceReport getDailyAttendanceReport(LocalDate attendanceDate);

    /**
     * Get attendance report for a specific session
     *
     * @param sessionId the ID of the course session
     * @return SessionAttendanceReport containing session attendance statistics
     */
    SessionAttendanceReport getSessionAttendanceReport(Integer sessionId);


    // ==================== Batch Operations ====================

    /**
     * Batch create multiple trainee attendance records
     *
     * @param attendances list of attendance data
     * @return List of NewRecordVTO containing created record IDs
     */
    List<NewRecordVTO> batchCreateTraineeAttendances(List<TraineeAttendanceDTO> attendances);


    // ==================== Lookup Operations ====================

    /**
     * Get all trainee attendance statuses for lookup
     *
     * @return LookupResultSet containing attendance statuses
     */
    LookupResultSet getTraineeAttendanceStatusLookup();
}

File: olympic-academy-working/src/app/core/services/trainee-attendance.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';
import { environment } from '../../../environments/environment';
import {
  TraineeAttendanceDTO,
  TraineeAttendanceResultSet,
  TraineeAttendanceVTO,
  DailyAttendanceReport,
  SessionAttendanceReport
} from '../models/trainee-attendance.model';
import { LookupResultSet } from '../../core/models/common.model';

@Injectable({
  providedIn: 'root'
})
export class TraineeAttendanceService {
  private apiUrl = `${environment.apiUrl}`;

  constructor(private http: HttpClient) {}

  // الحصول على حالات الحضور للـ Lookup
  getAttendanceStatusLookup(): Observable<LookupResultSet> {
    return this.http.get<LookupResultSet>(`${this.apiUrl}/lookup/trainee/attendance/status`);
  }

  // إنشاء سجل حضور
  createAttendance(data: TraineeAttendanceDTO): Observable<{ id: number }> {
    return this.http.post<{ id: number }>(`${this.apiUrl}/trainee-attendances`, data);
  }

  // إنشاء سجلات حضور جماعية
  batchCreateAttendance(data: TraineeAttendanceDTO[]): Observable<{ id: number }[]> {
    return this.http.post<{ id: number }[]>(`${this.apiUrl}/trainee-attendances/batch`, data);
  }

  // الحصول على جميع سجلات الحضور مع فلترة
  getAllAttendances(params?: {
    traineeId?: number;
    courseId?: number;
    courseSessionId?: number;
    status?: number;
    checkInFrom?: string;
    checkInTo?: string;
    checkOutFrom?: string;
    checkOutTo?: string;
    fromDate?: string;
    toDate?: string;
    pageNum?: number;
    pageSize?: number;
    orderDir?: string;
    orderBy?: string;
  }): Observable<TraineeAttendanceResultSet> {
    let httpParams = new HttpParams();
    if (params) {
      Object.keys(params).forEach(key => {
        const value = params[key as keyof typeof params];
        if (value !== null && value !== undefined) {
          httpParams = httpParams.set(key, value.toString());
        }
      });
    }
    return this.http.get<TraineeAttendanceResultSet>(`${this.apiUrl}/trainee-attendances`, { params: httpParams });
  }

  // الحصول على سجل حضور بالمعرف
  getAttendanceById(id: number): Observable<TraineeAttendanceVTO> {
    return this.http.get<TraineeAttendanceVTO>(`${this.apiUrl}/trainee-attendances/${id}`);
  }

  // تحديث سجل حضور
  updateAttendance(id: number, data: TraineeAttendanceDTO): Observable<{ id: number }> {
    return this.http.put<{ id: number }>(`${this.apiUrl}/trainee-attendances/${id}`, data);
  }

  // حذف سجل حضور
  deleteAttendance(id: number): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/trainee-attendances/${id}`);
  }

  // تقرير الحضور اليومي
  getDailyAttendanceReport(attendanceDate: string): Observable<DailyAttendanceReport> {
    return this.http.get<DailyAttendanceReport>(`${this.apiUrl}/trainees/attendances/reports/daily`, {
      params: { attendanceDate }
    });
  }

  // تقرير حضور جلسة محددة
  getSessionAttendanceReport(sessionId: number): Observable<SessionAttendanceReport> {
    return this.http.get<SessionAttendanceReport>(`${this.apiUrl}/trainee-attendances/session/${sessionId}/report`);
  }
}